tools/ledger-tool/src/main.rs

```diff
@@ -23,151 +17,23 @@ struct Args {
-        }
-        if !vesting_increment.contains('.') {
-            vesting_increment.extend(".000000000".chars());
-        }
-
-        account.timing = Timing::Timed {
-            initial_minimum_balance: Balance::of_mina_string_exn(&initial_minimum_balance),
-            cliff_time: Slot::from_u32(cliff_time[1].parse()?),
-            cliff_amount: Balance::of_mina_string_exn(&cliff_amount).to_amount(),
-            vesting_period: SlotSpan::from_u32(vesting_period[1].parse()?),
-            vesting_increment: Balance::of_mina_string_exn(&vesting_increment).to_amount(),
-        };
-    }
-    account_value.remove("sk");
-
-    if !account_value.is_empty() {
-        return Err(anyhow::anyhow!("the object contains unprocessed fields"));
-    }
-
-    Ok(account)
-}
-
 fn main() -> anyhow::Result<()> {
     let Args { input, url, output } = Args::from_args();
 
-    let value = if let Some(input) = input {
-        let ledger_file = File::open(&input)?;
-        serde_json::from_reader::<_, serde_json::Value>(ledger_file)?
+    let data = if let Some(input) = input {
+        fs::read(input)?
     } else if let Some(url) = url {
-        reqwest::blocking::get(url)?.json()?
+        reqwest::blocking::get(url)?.bytes()?.to_vec()
     } else {
         anyhow::bail!("must provide either `--input` or `--url`");
     };
 
-    let list = value
-        .as_object()
-        .ok_or_else(|| anyhow::anyhow!("expected object"))?
-        .get("ledger")
-        .ok_or_else(|| anyhow::anyhow!("expected field `ledger`"))?
-        .as_object()
-        .ok_or_else(|| anyhow::anyhow!("expected object"))?
-        .get("accounts")
-        .ok_or_else(|| anyhow::anyhow!("expected field `accounts`"))?
-        .as_array()
-        .ok_or_else(|| anyhow::anyhow!("expected array"))?;
-
-    let genesis_winner_account = {
-        let mut account = Account::empty();
-        account.public_key = CompressedPubKey::from_address(
-            "B62qiy32p8kAKnny8ZFwoMhYpBppM1DWVCqAPBYNcXnsAHhnfAAuXgg",
-        )
-        .expect("the constant is valid");
-
-        account.balance = Balance::of_nanomina_int_exn(1000);
-        account.delegate = Some(account.public_key.clone());
-        account
-    };
-
-    let mut accounts = vec![genesis_winner_account];
-    let mut mask = Mask::new_root(Database::create(35));
-
-    for (n, item) in list.iter().enumerate() {
-        accounts.push(
-            parse_account(item.clone())
-                .map_err(|err| anyhow::anyhow!("account: {n}, err: {err}"))?,
-        );
-    }
-
-    for account in &accounts {
-        let account_id = account.id();
-        mask.get_or_create_account(account_id, account.clone())
-            .expect("must not be full");
-    }
-
-    let root = mask.merkle_root();
-    let top_hash = v2::LedgerHash::from_fp(root);
-    println!("{top_hash}");
-
-    let mut hashes = vec![];
+    let daemon_json = serde_json::from_slice::<node::daemon_json::DaemonJson>(&data)?;
```

Comment:
Much better.

Btw the only reason we are doing this now is that the node doesn't cache the genesis ledger, but I think that is what should happen instead. And on startup, if the cache file is there (matched by hash?) it is loaded instead of parsing the config file to obtain the genesis ledger.

To speed up tests we can still have pre-generated files if necessary.

Reply:
The thing is that genesis ledger is not serializable. The closest thing is what we store in that prebuilt configuration.

Reply:
@akoptelov ok but in that case we can still serialize the same data we are producing with this tool and if it exists load that instead of parsing and processing the full config file.

Reply:
So, you mean if we use `--config ...`, we check if there's cached prebuilt configuration for it, and use it, otherwise generate and store it?

Reply:
Yes, at least for the accounts + hashes. Similar to what the OCaml node does with the genesis and epoch ledgers, which are stored in `~/.mina-config/genesis` and reused if they exists. This saves you from having to reconstruct the genesis ledger each time, and also from syncing up the epoch ledgers if you restart the node and the epoch is still the same as last time.

Reply:
You can see the stored ledgers (tagged by their hash):

```
> ls ~/.mina-config/genesis
epoch_ledger_3e2c138e3e014dcb66f5dcfbfd42e1c8cbbac7aeb5bbf90ed62907b91075a9f5
epoch_ledger_3e2c138e3e014dcb66f5dcfbfd42e1c8cbbac7aeb5bbf90ed62907b91075a9f5.tar.gz
epoch_ledger_b595fe2dbde04c83164b527df891ab445366a8dabe5ca448495e808a64c3ecdd
epoch_ledger_b595fe2dbde04c83164b527df891ab445366a8dabe5ca448495e808a64c3ecdd.tar.gz
genesis_ledger_7f8e3c72be57f296e86e1083379c7a09c6131b1567510cf0019eb4aff8b87813
genesis_ledger_7f8e3c72be57f296e86e1083379c7a09c6131b1567510cf0019eb4aff8b87813.tar.gz
```

Reply:
note that in OCaml they store the actual rocksdb database, but we don't need that, just accounts + hashes is enough to quickly reconstruct the ledger for us